File: art01.cpp
#include <iostream>
#include <unordered_map>
#include <string>
#include <vector>
#include <set>

unsigned long hash_djb2(const std::string& word) {
    unsigned long hash_value = 5381; // Inicialización del valor hash con un número primo
    for (char c : word) {
        hash_value = ((hash_value << 5) + hash_value) + c; // Aplicación de la función hash DJB2
    }
    return hash_value;
}

struct InvertedIndex {
    std::set<int> doc_ids;
};

int main() {
    std::vector<std::pair<int, std::string>> mapper_output = {
        {1, "Hello to all"},
        {2, "Hello World"},
        {3, "Love to all"},
        {4, "Love World"}
    }; // Entrada del mapper en el formato (doc_id, documento)

    std::unordered_map<std::string, std::set<int>> intermediate_output;

    // Mapper
    for (const auto& pair : mapper_output) {
        int doc_id = pair.first; // Obtener el doc_id del par
        std::string document = pair.second; // Obtener el documento del par
        std::string word;
        for (char c : document) {
            if (isspace(c)) { // Si el carácter es un espacio en blanco
                if (!word.empty()) { // Y la palabra actual no está vacía
                    intermediate_output[word].insert(doc_id); // Insertar la palabra y el doc_id en la salida intermedia
                    word.clear(); // Limpiar la palabra actual
                }
            } else {
                word += tolower(c); // Si el carácter no es un espacio, agregar el carácter en minúscula a la palabra actual
            }
        }
        if (!word.empty()) { // Si queda una palabra al final del documento
            intermediate_output[word].insert(doc_id); // Insertar la última palabra y el doc_id en la salida intermedia
        }
    }

    // Reducer
    const unsigned long table_size = intermediate_output.size(); // Calcular el tamaño de la tabla hash como el número de palabras únicas
    std::vector<std::pair<std::string, std::set<int>>> lookup_table(table_size); // Crear la tabla hash con el tamaño calculado

    for (const auto& pair : intermediate_output) {
        const std::string& word = pair.first; // Obtener la palabra del par
        const std::set<int>& doc_ids = pair.second; // Obtener el conjunto de doc_ids del par
        unsigned long position = hash_djb2(word) % table_size; // Calcular la posición hash de la palabra
        lookup_table[position] = std::make_pair(word, doc_ids); // Almacenar la palabra y el conjunto de doc_ids en la posición hash
    }

    // Output
    for (const auto& pair : lookup_table) {
        if (!pair.first.empty()) {
            std::cout << "(" << pair.first << ", ";
            for (int doc_id : pair.second) {
                std::cout << doc_id << " ";
            }
            std::cout << ")" << std::endl;
        }
    }

    return 0;
}

File: art01.py
def hash_djb2(word):
    hash_value = 5381  # Inicialización del valor hash con un número primo
    for c in word:
        hash_value = ((hash_value << 5) + hash_value) + ord(c)  # Aplicación de la función hash DJB2
    return hash_value


def main():
    # Entrada del mapper en el formato (doc_id, documento)
    mapper_output = [
        (1, "Hello to all"),
        (2, "Hello World"),
        (3, "Love to all"),
        (4, "Love World")
    ]

    # Diccionario para la salida intermedia del mapper
    intermediate_output = dict()

    # Mapper
    for pair in mapper_output:
        doc_id = pair[0]  # Obtener el doc_id del par
        document = pair[1]  # Obtener el documento del par

        word = ''  # Inicializar palabra actual

        for c in document:
            if c.isspace():  # Si el carácter es un espacio en blanco
                if word != '':  # Y la palabra actual no está vacía
                    word_lower = word.lower()  # Convertir a minúsculas
                    if word_lower not in intermediate_output:
                        intermediate_output[word_lower] = set()
                    intermediate_output[word_lower].add(doc_id)  # Insertar el doc_id
                    word = ''  # Limpiar la palabra actual
            else:
                word += c.lower()  # Agregar el carácter en minúscula

        if word != '':  # Si queda una palabra al final del documento
            word_lower = word.lower()
            if word_lower not in intermediate_output:
                intermediate_output[word_lower] = set()
            intermediate_output[word_lower].add(doc_id)

    # Reducer
    table_size = len(intermediate_output)  # Calcular el tamaño de la tabla hash
    lookup_table = []  # Crear la tabla de búsqueda como una lista

    for i in range(table_size):
        lookup_table.append(('', set()))  # Inicializar con tuplas vacías

    for pair in intermediate_output.items():
        word = pair[0]  # Obtener la palabra
        doc_ids = pair[1]  # Obtener el conjunto de doc_ids

        position = hash_djb2(word) % table_size  # Calcular la posición hash
        lookup_table[position] = (word, doc_ids)  # Almacenar en la tabla hash

    # Output
    for pair in lookup_table:
        word = pair[0]
        doc_ids = pair[1]
        if word != '':
            print("(", end='')
            print(word, end=', ')
            for doc_id in sorted(doc_ids):
                print(doc_id, end=' ')
            print(")")

if __name__ == "__main__":
    main()